### PR TITLE
Cdr 1613 implement transactions

### PR DESCRIPTION
**Pull request by harringj**

### CODE REVIEW

**Comment by bbpennel** [comment]
Actually it also looks like a bunch of tests are failing.  I believe that some if it is caused by not setting baseUri, like in the case of RepositoryTest.

TransactionalFcrepoClientTest is failing because it is calling `TransactionalFcrepoClient.client()` instead of `TransactionalFcrepoClient.client(baseUri)`.

I would suggest running all the tests in the FcrepoClient module

File: fcrepo-clients/src/test/java/edu/unc/lib/dl/fcrepo4/FedoraTransactionIT.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+		FolderObject obj = repository.createFolderObject(pid, model);
+
+		assertTrue(FedoraTransaction.hasTxId());
+		assertTrue(obj.getTypes().contains(Cdr.Folder.getURI()));
+		assertTrue(obj.getTypes().contains(PcdmModels.Object.getURI()));
+		assertEquals("Folder Title", obj.getResource()
+				.getProperty(DcElements.title).getString());
+		tx.close();
+		assertFalse(FedoraTransaction.hasTxId());
+		assertNull(tx.getTxUri());
```
Comment (harringj) [inline fcrepo-clients/src/test/java/edu/unc/lib/dl/fcrepo4/FedoraTransactionIT.java]: Good idea. Fixed!

File: fcrepo-clients/src/test/java/edu/unc/lib/dl/fcrepo4/FedoraTransactionIT.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+	}
+	
+	@Test (expected = TransactionCancelledException.class)
+	public void createRollbackTxTest() {
+		FedoraTransaction tx = repository.startTransaction();
+		repository.createFolderObject(pid, model);
+		tx.cancel();
```
Comment (harringj) [inline fcrepo-clients/src/test/java/edu/unc/lib/dl/fcrepo4/FedoraTransactionIT.java]: Done!

File: fcrepo-clients/src/test/java/edu/unc/lib/dl/fcrepo4/TransactionalFcrepoClientTest.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+import edu.unc.lib.dl.fcrepo4.TransactionalFcrepoClient.TransactionalFcrepoClientBuilder;
+
+public class TransactionalFcrepoClientTest extends AbstractFedoraTest {
+	
+	private static final String TX_URI = "http://localhost:48085/rest/tx:99b58d30-06f5-477b-a44c-d614a9049d38";
+	private static final String RESC_URI = "http://localhost:48085/rest/some/resource/id";
+	private static final String REQUEST_URI =
+			"http://localhost:48085/rest/tx:99b58d30-06f5-477b-a44c-d614a9049d38/some/resource/id";
+	
+	private TransactionalFcrepoClientBuilder builder = (TransactionalFcrepoClientBuilder) TransactionalFcrepoClient.client();
```
Comment (harringj) [inline fcrepo-clients/src/test/java/edu/unc/lib/dl/fcrepo4/TransactionalFcrepoClientTest.java]: Done!

File: fcrepo-clients/src/test/java/edu/unc/lib/dl/fcrepo4/FedoraTransactionIT.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+	public void nestedTxTest() throws Exception {
+		FedoraTransaction parentTx = repository.startTransaction();
+		repository.createFolderObject(pid, model);
+		
+		FedoraTransaction subTx = repository.startTransaction();
+		PID workPid = repository.mintContentPid();
+		repository.createWorkObject(workPid);
+		subTx.close();
+		
+		assertNull(subTx.getTxUri());
```
Comment (harringj) [inline fcrepo-clients/src/test/java/edu/unc/lib/dl/fcrepo4/FedoraTransactionIT.java]: Done! Like the idea of the helper, too :-)

**Pull request merged**